### PR TITLE
Parallel event loader components

### LINKED ISSUE #20438: Miscellaneous components for parallel event loader
Part of #20344. Add some independent loader components,

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Add some key components for the parallel event loader. Evidently it is incomplete (pending parser integration), but that will be completed in a follow-up.

**To test:**

Code review. This is not integrated anywhere yet so nothing can break.

Fixes #20438.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Parallel/inc/MantidParallel/IO/NXEventDataSource.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+class NXEventDataSource {
+public:
+  virtual ~NXEventDataSource() = default;
+
+  // Construct NXEVentDataLoader with list of bank names
```
Comment (LamarMoore) [inline Framework/Parallel/inc/MantidParallel/IO/NXEventDataSource.h]: This comment is in the wrong place

File: Framework/Parallel/src/IO/Chunker.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+namespace Mantid {
+namespace Parallel {
+namespace IO {
+
+namespace {
+/// Helper to build partition (subgroup of ranks with subgroup of banks).
+std::pair<int, std::vector<size_t>>
+buildPartition(const int totalWorkers, const size_t totalSize,
+               std::vector<std::tuple<size_t, size_t, bool>> &sortedSizes) {
```
Comment (OwenArnold) [inline Framework/Parallel/src/IO/Chunker.cpp]: I appreciate that this is a internal linkage free function, but a little hard to understand what for example sortedSizes type without looking at rest of code. I'm not suggesting changing the data structures at all.

File: Framework/Parallel/src/IO/Chunker.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+ * - Groups of workers and distribution of tasks to groups tries to balance
+ *   work, such that each worker has a roughly similar amount of work.
+ * Note that this method is public and static to allow for testing. */
+std::vector<std::pair<int, std::vector<size_t>>>
+Chunker::makeBalancedPartitioning(const int workers,
+                                  const std::vector<size_t> &sizes) {
+  const auto totalSize =
+      std::accumulate(sizes.begin(), sizes.end(), static_cast<size_t>(0));
+  // Indexed size vector such that we can sort it but still know original index.
+  // Elements are <size, original index, done flag>
```
Comment (OwenArnold) [inline Framework/Parallel/src/IO/Chunker.cpp]: can you also put this comment in the buildPartition method doxygen. I think that will solve my aforementioned comment.

File: Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+void load(
+    const Chunker &chunker,
+    NXEventDataSource<IndexType, TimeZeroType, TimeOffsetType> &dataSource) {
+  const size_t chunkSize = chunker.chunkSize();
+  const auto &ranges = chunker.makeLoadRanges();
+  // const auto &rankGroups = chunker.makeRankGroups();
+  std::vector<int32_t> event_id(2 * chunkSize);
+  std::vector<TimeOffsetType> event_time_offset(2 * chunkSize);
+
+  // TODO Create parser. Constructor arguments:
```
Comment (OwenArnold) [inline Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h]: When is TODO to be implemented?

File: Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
+#ifndef MANTID_PARALLEL_EVENTLOADERHELPERS_H_
```
Comment (OwenArnold) [inline Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h]: Quite a few TODOs in here. Is that deliberate?

File: Framework/Parallel/test/EventLoaderTest.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        load(H5::PredType::NATIVE_CHAR), const std::runtime_error &e,
+        std::string(e.what()),
+        "Unsupported H5::DataType for entry in NXevent_data");
+  }
+
+  void test_load() {
+    const std::vector<size_t> bankSizes{10, 100, 1000};
+    const size_t chunkSize{37};
+    Chunker chunker(1, 0, bankSizes, chunkSize);
+    ::detail::FakeDataSource dataSource;
```
Comment (OwenArnold) [inline Framework/Parallel/test/EventLoaderTest.h]: 👍  for arranging load methods in such a way as not to need file for  unittesting of this.

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Nicely done and neatly implemented. Very good attention to testing. A few issues around comments, but nothing else major.

File: Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h (hunk 5 above)
Comment (OwenArnold) [reply Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h]: Re-reading the PR description I'm thinking that this comes with the parser.

File: Framework/Parallel/src/IO/Chunker.cpp (hunk 2 above)
Comment (OwenArnold) [reply Framework/Parallel/src/IO/Chunker.cpp]: solution suggested below

File: Framework/Parallel/src/IO/Chunker.cpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+ * the task would be loading/processing a certain number of chunks from a file.
+ * The `workers` argument gives the number of workers to be used to process all
+ * tasks. The returned partitioning fulfils the following:
+ * - A task may be shared among workers. If workers share a task, all tasks they
+ *   are working on are shared among that group of workers.
+ * - Groups of workers and distribution of tasks to groups tries to balance
+ *   work, such that each worker has a roughly similar amount of work.
+ * Note that this method is public and static to allow for testing. */
+std::vector<std::pair<int, std::vector<size_t>>>
+Chunker::makeBalancedPartitioning(const int workers,
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply Framework/Parallel/src/IO/Chunker.cpp]: Well, depends on how it is used. In general this class should not care too much about whether the client pass in the actual number of ranks, or maybe something smaller if loading is to be done on just a subset of ranks. But I agree that I should maybe remove the term `rank` in some other places then.

File: Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h (hunk 5 above)
Comment (SimonHeybrock) [reply Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h]: Yes.

File: Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h (hunk 4 above)
Comment (SimonHeybrock) [reply Framework/Parallel/inc/MantidParallel/IO/EventLoaderHelpers.h]: https://github.com/mantidproject/mantid/tree/20887_integrate_parallel_event_loader

**Review by OwenArnold — APPROVED** [review]

**Comment by OwenArnold** [comment]
Happy for this to be merged once minor naming issue fixed (causing current compiler error failure) and builds pass.

**Pull request merged**